### ISSUE TITLE
Broken trace when recovering persistent actor in sharded cluster

### ISSUE THREAD

**Issue by garethjames-imburse** [petabridge/phobos-issues#31]
We've noticed an issue with Phobos tracing in our sharded cluster when a persistent child actor is started as a command is sent to it. The tracing appears to be working as expected until the child actor starts:

```
akka.msg.recv_MyActorShardEnvelope
 - akka.msg.recv_MyActorShardEnvelope
    - akka.actor.start

8<-------- break in tracing --------

akka.msg.recv_RequestRecoveryPermit
```

### akka.msg.recv_ShardEnvelope
```
akka.actor.path         = /system/sharding/myactor
akka.actor.recv.msgType = MyContracts.MyShardEnvelope
akka.actor.recv.sender  = akka.tcp://myactorsystem@myapi:8081/temp/vu
akka.actor.type         = Akka.Cluster.Sharding.ShardRegion
```

### akka.msg.recv_ShardEnvelope
```
akka.actor.path         = /system/sharding/myactor/myactor-id
akka.actor.recv.msgType = MyContracts.MyShardEnvelope
akka.actor.recv.sender  = akka.tcp://myactorsystem@myapi:8081/temp/vu
akka.actor.type         = Akka.Cluster.Sharding.Shard
```

> *MyActor receives the message **MyShardEnvelope** which has the real command inside and sends that command to MyChildActor:*

### akka.actor.start
```
akka.actor.path         = akka://myactorsystem/system/sharding/myactor/myactor-id/mychildactor-id#123456789
akka.actor.type         = MyApplication.MyChildActor
```

> *The trace appears to break here.*

### akka.msg.recv_RequestRecoveryPermit
```
akka.actor.path         = /system/recoveryPermitter
akka.actor.recv.msgType = Akka.Persistence.RequestRecoveryPermit
akka.actor.recv.sender  = akka://myactorsystem/system/sharding/myactor/myactor-id/mychildactor-id/$a
akka.actor.type         = Akka.Persistence.RecoveryPermitter
```

**Comment by Aaronontheweb** [comment]
Hi Gareth,

I have an open issue for looking into Akka.Persistence in general with Phobos - it looks to us like we don't tie the `Persist` before and after in together correctly. Do you have a timeframe you need this done by? We were going to work on that once we have our OTel tracing beta all the way out the door this month.

**Comment by garethjames-imburse** [comment]
Hi Aaron,

I wasn't sure if it was the same issue or a different one (albeit related), so I raised it separately. While we were waiting for an update, we put in a workaround for the other persistence issue which is to wrap the call to Persist and short circuit the trace manually as follows:
```csharp
protected void PersistWithTracing<TEvent>(TEvent @event, Action<TEvent> handler)
{
    if (@event is null)
        throw new ArgumentNullException(nameof(@event));

    var parentSpan = Context.GetInstrumentation().ActiveSpan;

    Persist(@event, _ =>
    {
        if (handler is null)
            return;

        var phobosActorContext = Context.GetInstrumentation();

        var spanBuilder = phobosActorContext.Tracer
            .BuildSpan($"custom.akka.persistence.handler_{@event.GetType().Name}")
            .AsChildOf(parentSpan);

        // Get tags using reflection
        var tags = GetTags(phobosActorContext.ActiveSpan);

        foreach (var tag in tags)
            spanBuilder.WithTag(tag.Key, tag.Value);

        using (spanBuilder.StartActive())
            handler(@event);
    });
}
```

In some cases this is working, but in other it's not and the trace in the description is one example where it isn't. If this is all related though I'm happy for you to roll the two issues up and close one; however you best want to track it.

As to a timeframe, I'll discuss it with the team and get back to you as soon as I have some info on that.

Many thanks!

**Comment by Aaronontheweb** [comment]
Progress update on our end: we've added some prototype instrumentation to capture all actor startup / restart activity prior to processing messages. This is a necessary ingredient for being able to trace recoveries from the point of view of the `PersistentActor`. The next stage is making sure that the persistence plugins themselves can properly correlate the activity on their end - this might require an update to Akka.Persistence itself to make the plugins more traceable, but we'll see.

**Comment by garethjames-imburse** [comment]
Thanks Aaron; appreciate the update!

**Comment by Aaronontheweb** [comment]
Whoops, not closed yet. Just merged in an Akka.NET PR that is needed to implement this in Phobos.

**Comment by Aaronontheweb** [comment]
Resolved via [Phobos 1.4.0](https://phobos.petabridge.com/articles/releases/RELEASE_NOTES.html#140httpssdkbincompublisherpetabridgeproductphobospackagesphobosactorversions140-january-19th-2022)
